xapian-core/common/str.cc

```diff
@@ -60,16 +60,23 @@ tostring(T value)
     if (value < 10 && value >= 0) return string(1, '0' + char(value));
 
     bool negative = (value < 0);
-    if (negative) value = -value;
 
-    char buf[(sizeof(T) * 5 + 1) / 2 + 1];
+    typedef typename std::make_unsigned<T>::type unsigned_type;
+    unsigned_type val;
+    if (negative) {
+	val = -value;
```

Comment:
If passed `int(-2147483648)` this will still negate `-2147483648` as an `int` (and only then convert the result to `unsigned int`).

Running `unittest` when in a build using ubsan I see:

```
Running test: tostring1...../common/str.cc:67:8: runtime error: negation of -2147483648 cannot be represented in type 'int'; cast to an unsigned type to negate this value to itself
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../common/str.cc:67:8 in 
```

I think the message there is suggesting doing this, which indeed seems to work:

```
    unsigned_type val(value);
    if (negative) {
       val = -val;
    }
```

https://en.cppreference.com/w/cpp/language/operator_arithmetic#Unary_arithmetic_operators explains the semantics of unary minus on an unsigned type.